File: app/src/main/java/at/tomtasche/reader/background/StreamUtil.java
package at.tomtasche.reader.background;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

public class StreamUtil {

    public static final String ENCODING = "UTF-8";

    public static void copy(File src, File dst) throws IOException {
        InputStream in = new FileInputStream(src);
        copy(in, dst);
    }

    public static void copy(File src, OutputStream out) throws IOException {
        InputStream in = new FileInputStream(src);
        copy(in, out);
    }

    public static void copy(InputStream in, File dst) throws IOException {
        OutputStream out = new FileOutputStream(dst);
        copy(in, out);
        out.close();
    }

    // taken from: https://stackoverflow.com/a/9293885/198996
    public static void copy(InputStream in, OutputStream out) throws IOException {
        try {
            // Transfer bytes from in to out
            byte[] buf = new byte[1024];
            int len;
            while ((len = in.read(buf)) > 0) {
                out.write(buf, 0, len);
            }

            out.flush();
        } finally {
            in.close();
        }
    }
}

File: app/src/main/java/app/opendocument/droid/background/StreamUtil.kt
package app.opendocument.droid.background

import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream

/** @JvmStatic throughout, because the java callers still spell these as static calls. */
object StreamUtil {

    const val ENCODING: String = "UTF-8"

    @JvmStatic
    @Throws(IOException::class)
    fun copy(src: File, dst: File) {
        copy(FileInputStream(src), dst)
    }

    @JvmStatic
    @Throws(IOException::class)
    fun copy(src: File, out: OutputStream) {
        copy(FileInputStream(src), out)
    }

    @JvmStatic
    @Throws(IOException::class)
    fun copy(input: InputStream, dst: File) {
        FileOutputStream(dst).use { out -> copy(input, out) }
    }

    @JvmStatic
    @Throws(IOException::class)
    fun readFully(input: InputStream): String {
        val out = ByteArrayOutputStream()
        copy(input, out)

        return out.toString(ENCODING)
    }

    /** Closes the input, never the output - the caller owns that one. */
    @JvmStatic
    @Throws(IOException::class)
    fun copy(input: InputStream, out: OutputStream) {
        input.use {
            it.copyTo(out)
            out.flush()
        }
    }
}
